File: src/old/bee.py
import asyncio


class Drone:
  def __init__(self, name, x, y, z) -> None:
    self.name = name
    self.x = x
    self.y = y
    self.z = z
    self.wayPoints = []

  async def init(self):
    print(self.name, "Ready")
    await self.takeOff()

  async def takeOff(self):
    print(self.name, "Taking off")
    await asyncio.sleep(1)
    print(self.name, "Airborne")
    return

  def takeWayPoints(self, givenWayPoints):
    self.wayPoints += givenWayPoints

  async def flyToWayPoints(self):
    if not len(self.wayPoints):
      print("No more wayPoints")
      return

    wayPoint, *self.wayPoints = self.wayPoints

    print(self.name, "Flying to ", wayPoint)

    await asyncio.sleep(5)

    self.x, self.y, self.z = wayPoint

    print(self.name, "Got to ", *wayPoint)

    await self.flyToWayPoints()


async def main():
  print('run main')

  wayPoints = [
      [1, 1, 1],
      [2, 2, 2],
      [3, 3, 3],
  ]

  myDrone = Drone("drone1", 0, 0, 0)
  myDrone.takeWayPoints(wayPoints)

  await myDrone.init()
  # create a task (its like create the promise)
  # the task is to fly to way points
  task = asyncio.create_task(
      myDrone.flyToWayPoints()
  )
  await asyncio.sleep(1)
  myDrone.takeWayPoints([[4, 4, 4]])

  # dont finish execution of the main function
  # until the task is done (no more points to go to)
  await task


asyncio.run(main())

# pending = asyncio.all_tasks()
# asyncio.loop.run_until_complete(asyncio.gather(*pending))

File: src/old/bee.js
"use scrict";


 

class Drone {
  constructor(name, x, y, z) {
    this.name = name;
    this.x = x;
    this.y = y;
    this.z = z;
    this.waypoints = []
  }

  async init() {
    console.log(this.name, "Ready");
    await this.takeOff();
  }

  async takeOff() {
    console.log(this.name, "Taking off");
    await new Promise(resolve => setTimeout(resolve, 1000)); //sleep 
    console.log(this.name, "Airborne");
    return
  }

  takeWaypoints(givenWaypoints = []) {
    this.waypoints = [...this.waypoints, ...givenWaypoints];
  }

  async flyToWaypoints() {
    if (this.waypoints.length === 0) {
      console.log("No more waypoints");
      return;
    };
    const waypoint = this.waypoints.shift();
    console.log(this.name, "Flying to", waypoint);
    await new Promise(resolve => setTimeout(resolve, 5000)); //sleep 
    this.x = waypoint[0];
    this.y = waypoint[1];
    this.z = waypoint[2];
    console.log(this.name, "Got to ", this.x, this.y, this.z);
    this.flyToWaypoints()
  }

};









(async ()=>{
  const myDrone0 = new Drone("drone1", 0, 0, 0)
  myDrone0.takeWaypoints([[1, 1, 1], [2, 2, 2], [3, 3, 3]]);
  await myDrone0.init();
  myDrone0.flyToWaypoints()
  setTimeout(()=>{
    myDrone0.takeWaypoints([ [4, 4, 4]]);
  },1000);  

})();



